File: algorithms/2_medium/P/00456_132_pattern/stack.cpp
class Solution {
public:
    static bool find132pattern(const vector<int>& nums) {
        auto stack = vector<int>();
        stack.reserve(nums.size());
        int third = numeric_limits<int>::min();

        for (auto it = nums.crbegin(); it != nums.crend(); it += 1) {
            const int num = *it;
            if (num < third) {
                return true;
            }

            while (!stack.empty() && num > stack.back()) {
                third = std::max(third, stack.back());
                stack.pop_back();
            }

            stack.push_back(num);
        }

        return false;
    }
};

File: algorithms/2_medium/P/00456_132_pattern/stack.rs
impl Solution {
    pub fn find132pattern(nums: Vec<i32>) -> bool {
        let mut stack: Vec<i32> = Vec::with_capacity(nums.len());
        let mut third = i32::MIN;

        for num in nums.into_iter().rev() {
            if num < third {
                return true;
            }

            while let Some(val) = stack.last() && num > *val {
                third = third.max(*val);
                stack.pop();
            }

            stack.push(num);
        }

        false
    }
}
